javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/InstrumentationModuleClassLoader.java

```diff
@@ -0,0 +1,246 @@
+ */
+
+package io.opentelemetry.javaagent.tooling.instrumentation.indy;
+
+import java.io.IOException;
+import java.lang.invoke.MethodHandles;
+import java.lang.reflect.Method;
+import java.net.URL;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.security.ProtectionDomain;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Classloader used to load the helper classes from {@link
+ * io.opentelemetry.javaagent.extension.instrumentation.InstrumentationModule}s, so that those
+ * classes have access to both the agent/extension classes and the instrumented application classes.
+ *
+ * <p>This classloader implements the following classloading delegation strategy:
+ *
+ * <ul>
+ *   <li>First, injected classes are considered (usually the helper classes from the
+ *       InstrumentationModule)
+ *   <li>Next, the classloader looks in the agent or extension classloader, depending on where the
+ *       InstrumentationModule comes from
+ *   <li>Finally, the instrumented application classloader is checked for the class
+ * </ul>
+ *
+ * <p>In addition, this classloader ensures that the lookup of corresponding .class resources follow
+ * the same delegation strategy, so that bytecode inspection tools work correctly.
+ */
+class InstrumentationModuleClassLoader extends ClassLoader {
+
+  private static final Map<String, ClassCopySource> ALWAYS_INJECTED_CLASSES =
+      Collections.singletonMap(
+          LookupExposer.class.getName(), ClassCopySource.create(LookupExposer.class).cached());
+  private static final ProtectionDomain PROTECTION_DOMAIN = getProtectionDomain();
+  private static final Method FIND_PACKAGE_METHOD = getFindPackageMethod();
+
+  private final Map<String, ClassCopySource> additionalInjectedClasses;
+  private final ClassLoader agentOrExtensionCl;
+  private final ClassLoader instrumentedCl;
+  private volatile MethodHandles.Lookup cachedLookup;
+
+  public InstrumentationModuleClassLoader(
+      ClassLoader instrumentedCl,
+      ClassLoader agentOrExtensionCl,
+      Map<String, ClassCopySource> injectedClasses) {
+    // agent/extension-classloader is "main"-parent, but class lookup is overridden
+    super(agentOrExtensionCl);
+    additionalInjectedClasses = injectedClasses;
+    this.agentOrExtensionCl = agentOrExtensionCl;
+    this.instrumentedCl = instrumentedCl;
+  }
+
+  /**
+   * Provides a Lookup within this classloader. See {@link LookupExposer} for the details.
+   *
+   * @return a lookup capable of accessing public types in this classloader
+   */
+  public MethodHandles.Lookup getLookup() {
+    if (cachedLookup == null) {
+      // Load the injected copy of LookupExposer and invoke it
+      try {
+        // we don't mind the race condition causing the initialization to run multiple times here
+        Class<?> lookupExposer = loadClass(LookupExposer.class.getName());
```

Comment:
Won't using the `LookupExposer.class` class load the class itself (presumably with the agent CL)? 
I.e. isn't this the same as the whole [advice class name problem](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/writing-instrumentation-module.md#define-the-actual-code-transformations-with-the-transformtypetransformer-method)?

Reply:
Yes, using `LookupExposer.class` will load the class in the agent-CL. This is however not a problem:
 * In contrast to Advices, `LookupExposer` only uses `java.`-classes and therefore can be loaded by the agent CL
 * The `InstrumentationModuleCL` is self-first, therefore its own copy of the class will be loaded here.
 
The downside is that we have to pay for the runtime overhead and memory consumption of unnecessarily loading the `LookupExposer` in the agent CL. The benefit of using `LookupExposer.class` is easier code navigation and better refactoring support. Imo, the runtime overhead is neglectable and worth to be paid.

I don't have a strong opinion here. If you prefer it we can replace the `LookupExposer.class` references with string literals.

Reply:
> The downside is that we have to pay for the runtime overhead and memory consumption of unnecessarily loading the `LookupExposer` in the agent CL. The benefit of using `LookupExposer.class` is easier code navigation and better refactoring support. Imo, the runtime overhead is neglectable and worth to be paid.

Yeah, it's just 1 more class reference. I'm fine with leaving this as it is 👍